MINOR: Improve the CI to avoid running long time tests when only document is changed

jiafu1115 opened the pull request:
if we only change md/html/etc. we don't need to run the unit test. it
will take hours to wait the final status sometimes

jiafu1115 committed:
improve the ci: avoid run the long time test if only document changed

Signed-off-by: Jian <fujian1115@gmail.com>

chia7712 commented:
@mumrah do you plan to add tests specifically for the documentation changes?

jiafu1115 commented:
If the documents have special check or test or have plan for it ，this pr is not needed

chia7712 commented:
Perhaps we can keep the check but skip test

jiafu1115 committed:
only filter test

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Update build.yml

jiafu1115 commented:
> Perhaps we can keep the check but skip test

@chia7712  done with change. to be honest the code is generated by AI with my review. It needs test with this CI and another CI with document change only. you can help to check if the approach make sense now. thanks a lot.

jiafu1115 committed:
Update build.yml

jiafu1115 committed:
remove useless code

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Merge remote-tracking branch 'origin/ciimprove' into ciimprove

Pankraz76 commented on .github/workflows/build.yml:
```diff
@@ -380,7 +400,7 @@ jobs:
       - name: Check Dependency Outcomes
         run: |
-          if [[ "${{ needs.validate.result }}" == "success" && "${{ needs.test.result }}" == "success" ]]; then
+          if [[ "${{ needs.validate.result }}" == "success" && ("${{ needs.configure.outputs.docChangeOnly }}" == "true" || "${{ needs.test.result }}" == "success") ]]; then
```
```suggestion
          if [[ "${{ needs.validate.result }}" == "success" && (${{ needs.configure.outputs.docChangeOnly }} || "${{ needs.test.result }}" == "success") ]]; then
```

is it possible to to like above, using system check for the value?

Pankraz76 commented on .github/workflows/build.yml:
```diff
@@ -175,7 +195,7 @@ jobs:
 
   test:
     needs: [configure, validate, load-catalog]
-    if: ${{ ! needs.configure.outputs.is-draft }}
```
nice fix.

Pankraz76 commented on .github/workflows/build.yml:
```diff
@@ -38,6 +38,7 @@ jobs:
       is-draft: ${{ steps.check-draft-pr.outputs.is-draft }}
       test-catalog-days: ${{ steps.configure-outputs.outputs.days }}
       sha: ${{ steps.configure-outputs.outputs.sha }}
+      docChangeOnly: ${{ steps.filter.outputs.docs == 'true' && steps.filter.outputs.code == 'false' }}
```
```suggestion
      docChangeOnly: ${{ steps.filter.outputs.docs && !steps.filter.outputs.code }}
```

jiafu1115 committed:
Adopt the suggestion

Co-authored-by: Vincent Potuček <8830888+Pankraz76@users.noreply.github.com>

jiafu1115 committed:
Adopt the suggestion

Co-authored-by: Vincent Potuček <8830888+Pankraz76@users.noreply.github.com>

chia7712 commented on .github/workflows/build.yml:
```diff
@@ -59,9 +60,28 @@ jobs:
+          persist-credentials: false
+      - name: Filter changed files
+        id: filter
+        uses: dorny/paths-filter@v2
```
Per Apache policy, we need to use git hash instead of tag version
https://infra.apache.org/github-actions-policy.html

jiafu1115 replied:
@chia7712 thanks for your reminder. done with the change according to https://github.com/apache/infrastructure-actions/blob/135d8bf57bd3eccd6812ca34d8bad888e7e75e5b/actions.yml#L311C1-L312C1

jiafu1115 committed:
Update paths-filter action to specific commit hash
Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Update build.yml

Add YAML files to paths-filter in build workflow for temp test

Update doc-change-only condition in build.yml

Fix condition check for doc-change-only output.

Just for test

Update AdminClientWithPoliciesIntegrationTest.java

Fix syntax for doc-change-only output in build.yml

Update build.yml

jiafu1115 committed:
unify file

Signed-off-by: Jian <fujian1115@gmail.com>

jiafu1115 committed:
Update CustomQuotaCallbackTest.java

jiafu1115 committed:
Update CustomQuotaCallbackTest.java

jiafu1115 commented:
Test result with follow **tmp** configure (BTW: I already changed back to right value after test):
```
        filters: |
            codes:
              - '**'
              - '!**/*.yml'
              - '!**/*.md
```

### Test Case 1
Change two files:
build.yml and CustomQuotaCallbackTest.java
The result: 
Filter codes = true
  
Matching files:
  clients/clients-integration-tests/src/test/java/org/apache/kafka/server/quota/CustomQuotaCallbackTest.java [modified]


### Test Case 2
Change one files
build.yml
The result: 
Filter codes = false
  Matching files: none

jiafu1115 committed:
Update build.yml

mumrah reviewed (commented):
Ideally, we don't need this since the Gradle build cache should avoid running tests if no source files have changed. But we still don't have that working perfectly, so an explicit opt-out of tests seems okay for docs changes. 

We should still run the "Compile and Check (Merge Ref)" job since it checks for licenses.

jiafu1115 commented:
> Ideally, we don't need this since the Gradle build cache should avoid running tests if no source files have changed. But we still don't have that working perfectly, so an explicit opt-out of tests seems okay for docs changes.
> 
> We should still run the "Compile and Check (Merge Ref)" job since it checks for licenses.

@mumrah thanks for your feedback. at the last. adopt the @chia7712 's suggestion. just ignore the test for document change only. example:
https://github.com/apache/kafka/actions/runs/20434902642

"But we still don't have that working perfectly,"- That's maybe the root cause I see some document only take few time when some other take a very long time.

thanks.

mumrah commented:
> That's maybe the root cause

Yes, very likely. I have an explanation here https://github.com/apache/kafka/pull/18449#issuecomment-2594038906

Essentially, we checkout `${{ github.event.pull_request.head.sha }}` which is the HEAD of the PR branch rather than the special "merge ref". This means if a docs PR is far behind trunk, it will get mostly cache misses.

chia7712 commented:
> Ideally, we don't need this since the Gradle build cache should avoid running tests if no source files have changed.

That would be the dream :smile: I really hope we can make that happen one day

jiafu1115 commented:
@chia7712 @mumrah so. does this PR make sense to do?
the dream example: **stable within few time**:
<img width="1865" height="515" alt="image" src="https://github.com/user-attachments/assets/66fc6a54-47aa-4718-808b-6d8fb0858f66" />

jiafu1115 committed:
code cleanup

jiafu1115 committed:
Fix output condition for document change detection

jiafu1115 committed:
tmp test

Exclude YAML files from code change detection.

jiafu1115 committed:
Update build.yml

jiafu1115 committed:
Update build.yml

jiafu1115 committed:
Update build.yml

jiafu1115 committed:
Update AdminApiDriver.java

jiafu1115 committed:
Update AdminApiDriver.java

jiafu1115 committed:
Allow YAML files in code change detection

Removed exclusion for YAML files in code change detection.

jiafu1115 committed:
Update build.yml

jiafu1115 commented:
@mumrah @chia7712 @Pankraz76 

Thanks for your discussion. I just completed the code refactor and tests. I also add more file types which need to be exclude from tests after the CI test.

**It is the final status for me now. If you think it is worthy for this PR. you can help to review. Thanks.**

### Test 1: Only doc change
<img width="658" height="562" alt="image" src="https://github.com/user-attachments/assets/0322ebc8-b387-4ff9-8ddd-2b959569ee3c" />

### Test 2: Only code change
<img width="571" height="425" alt="image" src="https://github.com/user-attachments/assets/2b87e9e6-0368-4a0c-b143-e827f4a47656" />

### Test 3: Code and Doc changed
<img width="787" height="527" alt="image" src="https://github.com/user-attachments/assets/f6c84fae-b2b8-475d-8d01-bf1d9dda6e98" />

BTW. I record some information found in tests here for your reference:

1. "predicate-quantifier: 'every'" will got one warn log but it shouldn't be deleted due to the behavior will be changed accroding to the test result.

`_Warning: Unexpected input(s) 'predicate-quantifier', valid inputs are ['token', 'working-directory', 'ref', 'base', 'filters', 'list-files', 'initial-fetch-depth']_`

2. ${{ steps.filter.outputs.docs && !steps.filter.outputs.codes }}" **is not equal with** ${{ steps.filter.outputs.docs == 'true' && steps.filter.outputs.codes == 'false' }}". @Pankraz76  so I rollback part of the code. I also check the plugin example (https://github.com/dorny/paths-filter). you can check my test result here directly.
```
echo "steps.filter.outputs.docs: ${{ steps.filter.outputs.docs }}"
echo "steps.filter.outputs.codes: ${{ steps.filter.outputs.codes }}"
echo "doc-change-only: ${{ steps.filter.outputs.docs && !steps.filter.outputs.codes }}"
echo "doc-change-only: ${{ steps.filter.outputs.docs == 'true' && steps.filter.outputs.codes == 'false' }}"
the result: true. false. false, true
```
Thanks !

Regards
Jian

Pankraz76 commented on .github/workflows/build.yml:
```diff
@@ -175,7 +197,7 @@ jobs:
   test:
     needs: [configure, validate, load-catalog]
-    if: ${{ ! needs.configure.outputs.is-draft }}
+    if: ${{ !needs.configure.outputs.is-draft && needs.configure.outputs.doc-change-only != 'true' }}
```
```suggestion
    if: ${{ !needs.configure.outputs.is-draft && needs.configure.outputs.doc-change-only}}
```

imho you could use the native boolean interpretation.
You might need to negate, but you dont need to manually evaluate the boolean yourself.

just like seen in the EXP before. `!needs.configure.outputs.is-draft`

this feels like feature envy.

jiafu1115 replied:
@Pankraz76 thanks for your comment. It looks like no one is particularly interested in this PR or want this PR, so I’ll keep things as they are for now. That said, I really appreciate your comments—thanks a lot!
